### PR TITLE
fix(plugin-ioredis): end span on response from the server and set span status according to response

### PR DESCRIPTION
**Pull request by blumamir**
## Which problem is this PR solving?

- fix: end span on response from server, so duration accurately include the server processing and networking time. Currently, the span is ended after *sending* the command to the server.
- fix: set span status to error if server response with error. Currently, span status is always set is OK, even if the server replied with an error. 

## Short description of the changes
patch command `resolve` and `reject` functions. Only when the response arrives from the server - close the span (so the duration is accurate), and set status code to error if the server responds with error.

### CODE REVIEW

**Comment by dyladan** [comment]
> * feat: add responseHook to plugin config so user can register hook to add custom attributes to
>   span when successful response arrive from server

I would prefer if this was in a separate PR so it can be tracked in the changelog as a feature

**Comment by blumamir** [comment]
@dyladan @vmarchaud Done.
Will appreciate your review so I can create the next PR soon when this one merges into master

**Review by vmarchaud — APPROVED** [review]

**Comment by blumamir** [comment]
Thanks @vmarchaud 
Does this PR need to have a `bug` label so it shows up in the changelog?
What is the policy of merging PRs when they are approved?

**Comment by dyladan** [comment]
In contrib:

- 3 reviews required
- 1 review should be from a maintainer
- If the PR is opened by a maintainer, only 2 reviews are required
- No unresolved comments/conversations
- No changes requested
- Tests/lint pass

**Comment by dyladan** [comment]
Yes the bug label makes this show up in the changelog.

File: plugins/node/opentelemetry-plugin-ioredis/src/utils.ts
Comment (naseemkullah) [inline plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:38]: What are your thoughts of putting a ternary here instead of the let if before?

**Review by naseemkullah — APPROVED** [review]
Litmus, left a thought

File: plugins/node/opentelemetry-plugin-ioredis/src/utils.ts
Comment (blumamir) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts]: good idea. pushed an update

**Comment by obecny** [comment]
@blumamir pls fix the lint so I can merge it, thx

**Comment by blumamir** [comment]
The lint error is not related to my changes:
`ESLint couldn't find the config "./node_modules/gts" to extend from. Please check that the name of the config is correct.`

When running locally on ioredis package I get no lint errors, but when running on the entire repo, it fails due to `@opentelemetry/instrumentation-graphql`:

```
lerna ERR! npm run lint exited 1 in '@opentelemetry/instrumentation-graphql'
lerna ERR! npm run lint stdout:

> @opentelemetry/instrumentation-graphql@0.11.0 lint /Users/amirblum/repos/opentelemetry-js-contrib/plugins/node/opentelemetry-instrumentation-graphql
> eslint . --ext .ts


/Users/amirblum/repos/opentelemetry-js-contrib/plugins/node/opentelemetry-instrumentation-graphql/src/graphql.ts
   96:23  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  156:48  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  156:53  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  386:16  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  387:19  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  388:44  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  390:60  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  390:65  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  391:58  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  391:63  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

/Users/amirblum/repos/opentelemetry-js-contrib/plugins/node/opentelemetry-instrumentation-graphql/src/types.ts
   66:59  error    Replace `GraphQLInstrumentationConfig` with `⏎··GraphQLInstrumentationConfig⏎`  prettier/prettier
   76:3   warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   77:3   warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   78:26  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   80:43  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   80:48  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   81:42  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   81:47  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   87:15  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   88:18  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   89:43  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   91:59  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   91:64  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   92:57  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   92:62  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   98:15  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
   99:18  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
  100:43  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
  102:46  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
  102:51  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
  103:44  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
  103:49  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any
  128:12  warning  Unexpected any. Specify a different type                                        @typescript-eslint/no-explicit-any

/Users/amirblum/repos/opentelemetry-js-contrib/plugins/node/opentelemetry-instrumentation-graphql/src/utils.ts
   54:17  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
   58:10  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
   91:17  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  156:33  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  162:33  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  166:39  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  299:26  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  309:45  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  309:61  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  309:74  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  339:38  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  341:16  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

✖ 45 problems (1 error, 44 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.


lerna ERR! npm run lint stderr:
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @opentelemetry/instrumentation-graphql@0.11.0 lint: `eslint . --ext .ts`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @opentelemetry/instrumentation-graphql@0.11.0 lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

How can I make it pass?

**Comment by blumamir** [comment]
The lint passed until I merged master into this branch [here](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/239/commits/38a3818c7ab392d0bdf39a235e1d9efbd7d2661a)

**Comment by obecny** [comment]
@blumamir just run `npm run lint:fix` but make sure you have up to date all npm packages, obviously some small change in lint rules recently and it is failing on white space

**Comment by blumamir** [comment]
> @blumamir just run `npm run lint:fix` but make sure you have up to date all npm packages, obviously some small change in lint rules recently and it is failing on white space

@obecny 
I run it, and it changed a file not related to my PR: 
plugins/node/opentelemetry-instrumentation-graphql/src/types.ts

Is it valid?

**Comment by blumamir** [comment]
@obecny 
The lint check is still failing in the CI with this error: 
`ESLint couldn't find the config "./node_modules/gts" to extend from. Please check that the name of the config is correct.`

Locally it passes with no errors. not sure how to fix it

**Comment by obecny** [comment]
@blumamir weird, I can just suggest to do a full cleanup (when many packages changes, you have old package with lock etc. somethimes npm dosent update packages correctly). The easiest way is to remove all node_modules, then app package-lock.json, and all build folders, if you are on mac I have handy gist here to do it in one command you can just copy and paste in main package.json then run "npm run clean:all" and after this run "npm i"

https://gist.github.com/obecny/d4358fce87a862b6b3f6ba54398ac30e/e37abf6a171f33ae20684cff2ab4e1c3e513493a

**Comment by obecny** [comment]
a wait

**Comment by obecny** [comment]
seems like it might be the github actions , checking ....

**Comment by obecny** [comment]
@blumamir just merged the latest to my PR `https://github.com/open-telemetry/opentelemetry-js-contrib/pull/277` and it doesn't fail on lint. Also the modification you have in graphql is not needed. This might be some caching issue , not sure :/ @dyladan any idea if we can "reset" this PR to fresh start without getting cache ?

**Comment by vmarchaud** [comment]
@blumamir Could you rebase from master, i believe the previous lint error should be fixed ?

**Comment by blumamir** [comment]
> @blumamir Could you rebase from master, i believe the previous lint error should be fixed ?

@vmarchaud I updated the PR following the recent changes in StatusCode, and change to the test `'should create a child span for lua'` which validated the function call response (OK -> StatusCode.UNSET) instead of the redis server response (NOSCRIPT -> StatusCode.ERROR).
But I still get lint errors on graphql.
Running `npm run lint` on `upstream\master` fails as well on my setup.
Does it works for you? seems like an issue unrelated to my change.

BTW, I see there is a PR for porting ioredis to the instrumentation API. Is there any point in merging this PR into master? I assume the `plugin` package will be removed once the `instrumentation` package will be merged.

**Comment by Flarna** [comment]
> BTW, I see there is a PR for porting ioredis to the instrumentation API. Is there any point in merging this PR into master?

In case this PR gets merged first I will update that one for the instrumentation package. I don't know if there is a concrete timeline when to remove the plugin packages.

Regarding the lint error: I think you have to undo your change in graphql. I had a similar issue in the past and it was caused by different versions of installed npm packages lying around on my PC compared to that ones installed fresh in CI. Try to delete all package-lock.json files and node_modules folders and then do a fresh npm install in root folder of the project.

**Comment by blumamir** [comment]
> Regarding the lint error: I think you have to undo your change in graphql. I had a similar issue in the past and it was caused by different versions of installed npm packages lying around on my PC compared to that ones installed fresh in CI. Try to delete all package-lock.json files and node_modules folders and then do a fresh npm install in root folder of the project.

Thanks @Flarna , that fixed the lint error :)
@vmarchaud can this PR be merged?

**Comment by vmarchaud** [comment]
> In case this PR gets merged first I will update that one for the instrumentation package. I don't know if there is a concrete timeline when to remove the plugin packages.

We didn't discuss this i believe, however i think we aim for all core plugins to be ported (still grpc-js left but PR is open)

> @vmarchaud can this PR be merged?

Just need one review from another maintainer /cc @obecny @dyladan

File: plugins/node/opentelemetry-plugin-ioredis/src/utils.ts
Comment (obecny) [inline plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:122]: Ideally we should use `recordException` inside endSpan function as `err` might contain information also about stack and code but I see there is used `setStatus`

Comment (blumamir) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:122]: I think it's not in the scope of this PR. I used the already existing function as it is.
Maybe you can open an issue for that?

Comment (obecny) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:122]: the description says `fix: set span status to error if server response with error. ` isn't this exactly connected with saving information about error ?

Comment (blumamir) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:122]: This PR is about setting the error according to the server response, instead of falsely setting it according to function call returned value. I prefer to keep it minimal and to fix a single issue.
I agree that there is room for other improvements to this plugin code as you mentioned.

Comment (blumamir) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:129]: Hi @obecny ,
Just checking about the status of this comment.
This PR is waiting for almost 3 months

Comment (blumamir) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:129]: @obecny 
Do we agree that there is currently a bug described above in this plugin code?
Do we agree that the PR will not introduce **new** bugs or issues to the code according to your review?
Do we agree that the change in this PR is fixing the above issue?

If so, I have to say that I don't understand why you block it from being merged.
Is it the official policy of this project?

Comment (obecny) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:129]: @blumamir two things

1. technical - `recordException` functionality is part of official spec and is implemented in js. If error occurs we should store the information about error so that it is easy to understand what has happened. Stack trace is one of the element that helps debugging and fixing stuff. This will definitely help user to understand what has happened. For this purpose `recordException` has been added and I would say it is highly recommended when you store the information that the span has error.
 
2. Blaming me for blocking this PR and that the ticket has been opened so long. 
If you look look closely what are the rules for merging you will see that even if I wasn't here this PR could be merged if it meets the merging criteria which is doable even without my approve. You can also see that I already said in December that this can be merged just fix the lint errors otherwise I cannot merge or any1 else when the build is broken. It wasn't fixed for quite long time was just there (or correct me if I'm wrong but then nothing was said to just attract some attention for more than month) until @vmarchaud asked you to fix the build again. Then last week I was asked again for one more review which I did and raised the issue that it would make sense to use also `recordException` (btw 30s job, one line of code - it will be longer to create a new issue then adding it here). But you didn't like it not sure why though. 

If I wanted to block this particular PR I would click "request changes" then it would be blocked. But I didn't do that, you just don't have my approve as I think it still lacks `recordException`. You could raise a separate issue for that so it won't get lost and any1 else can pick it up and add it in separate PR. But I don't understand why I have to approve something if I think it still lacks things (in my human opinion). 

Your last sentence is then the least understandable for when you say about policy of this project, but I'm not going to comment on that. 

And as last thing, this is quite a big project some people that were approvers are not longer active here. There are more work and less people. We have recently decided it to improve that too. Have a new maintainer and 2 new approvers - how it will work we will see in next couple weeks, but I think it is already a big improvement.

Comment (blumamir) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:129]: Thanks for the detailed response.
I appreciate the work and time you put into this project and didn't mean to blame you for anything. Sorry if it sounds like this. I thought your review blocks this PR from being merged, but I see now that it is not. I'll just wait for other reviews in that case.
I understand what you are saying about `recordException` and would love to see it get fixed as well, just don't want to mix it into this PR which is fixing something else.

Comment (naseemkullah) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:129]: My understanding is that this unresolved conversation indeed blocks merge (along with the lack of 2 maintainer approvals)  as per https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md#general-merge-requirements

> No unresolved conversations.

Which in itself is a questionable/hard to interpret requirement as the PR author could resolve the conversation themself.

Comment (obecny) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:129]: > My understanding is that this unresolved conversation indeed blocks merge (along with the lack of 2 maintainer approvals) as per https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md#general-merge-requirements

@naseemkullah You are right but until the PR receives enough approves it doesn't block anything. Once it receives all needed approves then all unresolved conversations should be resolved so that nothing hangs behind. It still didn't happen after so long time. As I mentioned before any of following 2 will work for me: either creating a new issue so it is not forgotten and can be done in separate PR or add recordException. Both ways works fine for me. But it doesn't stop other people from reviewing it, the more reviews the better.

Comment (blumamir) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:129]: @obecny I'll add the `recordException` tomorrow and update the PR.

Comment (blumamir) [reply plugins/node/opentelemetry-plugin-ioredis/src/utils.ts:130]: @obecny 
Added the `recordException` to the PR and updated the tests accordingly.

**Review by obecny — APPROVED** [review]
lgtm, thx for changes

**Comment by obecny** [comment]
@blumamir looks like tests for ioredis is broken, once fixed this can be merged

**Comment by blumamir** [comment]
> @blumamir looks like tests for ioredis is broken, once fixed this can be merged

I'll fix it now, thanks

**Comment by blumamir** [comment]
> @blumamir looks like tests for ioredis is broken, once fixed this can be merged

@obecny 
tests are fixed

> > BTW, I see there is a PR for porting ioredis to the instrumentation API. Is there any point in merging this PR into master?
> 
> In case this PR gets merged first I will update that one for the instrumentation package. I don't know if there is a concrete timeline when to remove the plugin packages.

@Flarna if I can assist somehow in merging this fix into the instrumentation PR, please let me know

**Comment by obecny** [comment]
will merge this one first. the plugins will still remain for some time until all is moved to instrumentation, then will have to make more changes in core and then they can be deprecated and then finally remove. It  will not happen in few days for sure.

**Comment by Flarna** [comment]
@blumamir I will copy over your changes into the instrumentation later today. In case I need some help I will ping you but it doesn't look that complicated.

**Pull request merged**
